src/account/PluginManager2.sol

```diff
@@ -6,63 +6,109 @@ import {EnumerableSet} from "@openzeppelin/contracts/utils/structs/EnumerableSet
 
 // Temporary additional functions for a user-controlled install flow for validation functions.
 abstract contract PluginManager2 {
     using EnumerableSet for EnumerableSet.Bytes32Set;
 
-    error DefaultValidationAlreadySet(address plugin, uint8 functionId);
-    error ValidationAlreadySet(bytes4 selector, address plugin, uint8 functionId);
-    error ValidationNotSet(bytes4 selector, address plugin, uint8 functionId);
+    error DefaultValidationAlreadySet(FunctionReference validationFunction);
+    error ValidationAlreadySet(bytes4 selector, FunctionReference validationFunction);
+    error PreValidationAlreadySet(FunctionReference validationFunction, FunctionReference preValidationFunction);
+    error ValidationNotSet(bytes4 selector, FunctionReference validationFunction);
 
     function _installValidation(
-        address plugin,
-        uint8 functionId,
+        FunctionReference validationFunction,
         bool isDefault,
         bytes4[] memory selectors,
-        bytes calldata installData
-    ) internal {
-        FunctionReference validationFunction = FunctionReferenceLib.pack(plugin, functionId);
-
+        bytes calldata installData,
+        bytes memory preValidationHooks
+    )
+        // TODO: flag for signature validation
+        internal
+    {
         AccountStorage storage _storage = getAccountStorage();
 
+        if (preValidationHooks.length > 0) {
+            (FunctionReference[] memory preValidationFunctions, bytes[] memory initDatas) =
+                abi.decode(preValidationHooks, (FunctionReference[], bytes[]));
+
+            for (uint256 i = 0; i < preValidationFunctions.length; ++i) {
+                FunctionReference preValidationFunction = preValidationFunctions[i];
+
+                if (
+                    !_storage.validationData[validationFunction].preValidationHooks.add(
+                        toSetValue(preValidationFunction)
+                    )
+                ) {
+                    revert PreValidationAlreadySet(validationFunction, preValidationFunction);
+                }
+
+                if (initDatas[i].length > 0) {
+                    (address preValidationPlugin,) = FunctionReferenceLib.unpack(preValidationFunction);
+                    IPlugin(preValidationPlugin).onInstall(initDatas[i]);
+                }
+            }
+        }
+
         if (isDefault) {
-            if (!_storage.defaultValidations.add(toSetValue(validationFunction))) {
-                revert DefaultValidationAlreadySet(plugin, functionId);
+            if (_storage.validationData[validationFunction].isDefault) {
+                revert DefaultValidationAlreadySet(validationFunction);
             }
+            _storage.validationData[validationFunction].isDefault = true;
         }
 
         for (uint256 i = 0; i < selectors.length; ++i) {
             bytes4 selector = selectors[i];
             if (!_storage.selectorData[selector].validations.add(toSetValue(validationFunction))) {
```

Comment:
I remember native functions can be validated by any default validation.
If the selector is a native function, and the validation function is default, can we skip this step?

Reply:
Yes, this step can be skipped in that case. For example, the new initialization method passes an empty `bytes4[] selectors` to this function: https://github.com/erc6900/reference-implementation/blob/4d2390be98f38fd71e1b28e7b0a52ea141088fe9/src/account/UpgradeableModularAccount.sol#L337